File: backend/src/controllers/employees.ts
import { Router } from "express";

import { PAGE_SIZE_DEFAULT, PAGE_SIZE_MAX } from "../common/constants.js";
import type { Employee, SalaryRecord } from "../common/types.js";
import type { EmployeeService } from "../services/employees.js";
import type { SalaryService } from "../services/salaries.js";
import { decodeExcelBase64, parseEmployeeExcel, sampleEmployeeExcel } from "../utils/excel.js";
import { queryParam, queryValues } from "../utils/query.js";
import { createEmployeeBody, importExcelBody, updateEmployeeBody } from "../validators/employees.js";
import { parseAmount, salaryBody } from "../validators/salaries.js";

function toSalaryJson(record: SalaryRecord) {
  return {
    id: record.id,
    employee_id: record.employeeId,
    base_amount: record.baseAmount.toFixed(2),
    currency: record.currency,
    effective_from: record.effectiveFrom,
    reason: record.reason,
    created_at: record.createdAt.toISOString(),
  };
}

function parsePositiveInt(value: unknown, fallback: number, max = Number.MAX_SAFE_INTEGER): number {
  const n = Number(value);
  if (!Number.isInteger(n) || n < 1) return fallback;
  return Math.min(max, n);
}

function toEmployeeJson(employee: Employee, currentSalary?: SalaryRecord | null) {
  return {
    id: employee.id,
    employee_code: employee.employeeCode,
    name: employee.name,
    email: employee.email,
    country_code: employee.countryCode,
    department: employee.department,
    designation: employee.designation,
    status: employee.status,
    created_at: employee.createdAt.toISOString(),
    updated_at: employee.updatedAt.toISOString(),
    current_salary: currentSalary ? toSalaryJson(currentSalary) : null,
  };
}

export function employeeRoutes(deps: {
  employees: EmployeeService;
  salaries: SalaryService;
  today: () => string;
}) {
  const router = Router();

  router.get("/", async (req, res, next) => {
    try {
      const page = parsePositiveInt(queryParam(req, "page"), 1);
      const pageSize = parsePositiveInt(queryParam(req, "page_size"), PAGE_SIZE_DEFAULT, PAGE_SIZE_MAX);
      const { items, total } = await deps.employees.listPage({
        names: queryValues(req.query.q),
        countries: queryValues(req.query.country),
        departments: queryValues(req.query.department),
        statuses: queryValues(req.query.status),
        offset: (page - 1) * pageSize,
        limit: pageSize,
      });
      const currentByEmployeeId = await deps.salaries.getCurrentForMany(
        items.map((row) => row.id),
        deps.today(),
      );
      res.json({
        items: items.map((row) => toEmployeeJson(row, currentByEmployeeId.get(row.id))),
        page,
        page_size: pageSize,
        total,
      });
    } catch (error) {
      next(error);
    }
  });

  router.post("/", async (req, res, next) => {
    try {
      const body = createEmployeeBody.parse(req.body);
      const employee = await deps.employees.create({
        name: body.name,
        email: body.email,
        countryCode: body.country_code,
        department: body.department,
        designation: body.designation,
        employeeCode: body.employee_code,
      });
      res.status(201).json(toEmployeeJson(employee, null));
    } catch (error) {
      next(error);
    }
  });

  router.get("/import/sample", async (_req, res, next) => {
    try {
      const file = await sampleEmployeeExcel();
      res.setHeader(
        "Content-Type",
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
      );
      res.setHeader("Content-Disposition", "attachment; filename=employees-sample.xlsx");
      res.send(file);
    } catch (error) {
      next(error);
    }
  });

  router.post("/import", async (req, res, next) => {
    try {
      const { file } = importExcelBody.parse(req.body);
      const { created, errors } = await deps.employees.createMany(
        await parseEmployeeExcel(decodeExcelBase64(file)),
      );
      res.json({ created: created.length, errors });
    } catch (error) {
      next(error);
    }
  });

  router.get("/:id", async (req, res, next) => {
    try {
      const employeeId = req.params.id;
      const employee = await deps.employees.getById(employeeId);
      const currentSalary = await deps.salaries.getCurrent(employee.id, deps.today());
      res.json(toEmployeeJson(employee, currentSalary));
    } catch (error) {
      next(error);
    }
  });

  router.patch("/:id", async (req, res, next) => {
    try {
      const body = updateEmployeeBody.parse(req.body);
      const employee = await deps.employees.updateById(req.params.id, {
        name: body.name,
        email: body.email,
        countryCode: body.country_code,
        department: body.department,
        designation: body.designation,
        status: body.status,
      });
      const currentSalary = await deps.salaries.getCurrent(employee.id, deps.today());
      res.json(toEmployeeJson(employee, currentSalary));
    } catch (error) {
      next(error);
    }
  });

  router.get("/:id/salary-history", async (req, res, next) => {
    try {
      const salaryHistory = await deps.salaries.listHistory(req.params.id);
      res.json(salaryHistory.map(toSalaryJson));
    } catch (error) {
      next(error);
    }
  });

  router.post("/:id/salary", async (req, res, next) => {
    try {
      const body = salaryBody.parse(req.body);
      const salary = await deps.salaries.addSalary({
        employeeId: req.params.id,
        baseAmount: parseAmount(body.base_amount),
        currency: body.currency,
        effectiveFrom: body.effective_from,
        reason: body.reason,
      });
      res.status(201).json(toSalaryJson(salary));
    } catch (error) {
      next(error);
    }
  });

  return router;
}

File: backend-py/app/routers/employees.py
from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, Query, Request
from fastapi.responses import JSONResponse, Response
from pydantic import BaseModel, Field

from app.constants import PAGE_SIZE_DEFAULT, PAGE_SIZE_MAX
from app.excel import decode_excel_base64, parse_employee_excel, sample_employee_excel
from app.jsonutil import employee_json, salary_json
from app.models import ListFilters
from app.query import parse_positive_int, query_values
from app.validators import parse_amount

router = APIRouter(prefix="/api/employees")


class CreateEmployeeBody(BaseModel):
    name: str
    email: str
    country_code: str
    department: str
    designation: str
    employee_code: str | None = None


class UpdateEmployeeBody(BaseModel):
    name: str | None = None
    email: str | None = None
    country_code: str | None = None
    department: str | None = None
    designation: str | None = None
    status: str | None = None


class ImportExcelBody(BaseModel):
    file: str = Field(min_length=1)


class SalaryBody(BaseModel):
    base_amount: str | int | float
    currency: str
    effective_from: str
    reason: str


def _deps(request: Request):
    return request.app.state.employees, request.app.state.salaries, request.app.state.today


@router.get("")
def list_employees(
    request: Request,
    page: str | None = None,
    page_size: str | None = None,
    q: Annotated[list[str] | None, Query()] = None,
    country: Annotated[list[str] | None, Query()] = None,
    department: Annotated[list[str] | None, Query()] = None,
    status: Annotated[list[str] | None, Query()] = None,
):
    employees, salaries, today = _deps(request)
    page_n = parse_positive_int(page, 1)
    size = parse_positive_int(page_size, PAGE_SIZE_DEFAULT, PAGE_SIZE_MAX)
    items, total = employees.list_page(
        ListFilters(
            names=query_values(q),
            countries=query_values(country),
            departments=query_values(department),
            statuses=query_values(status),
            offset=(page_n - 1) * size,
            limit=size,
        )
    )
    current = salaries.get_current_for_many([row.id for row in items], today())
    return {
        "items": [employee_json(row, current.get(row.id)) for row in items],
        "page": page_n,
        "page_size": size,
        "total": total,
    }


@router.post("", status_code=201)
def hire(request: Request, body: CreateEmployeeBody):
    employees, _, _ = _deps(request)
    employee = employees.create(
        name=body.name,
        email=body.email,
        country_code=body.country_code,
        department=body.department,
        designation=body.designation,
        employee_code=body.employee_code,
    )
    return employee_json(employee, None)


@router.get("/import/sample")
def import_sample():
    data = sample_employee_excel()
    return Response(
        content=data,
        media_type="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        headers={"Content-Disposition": "attachment; filename=employees-sample.xlsx"},
    )


@router.post("/import")
def import_employees(request: Request, body: ImportExcelBody):
    employees, _, _ = _deps(request)
    created, errors = employees.create_many(parse_employee_excel(decode_excel_base64(body.file)))
    return {"created": len(created), "errors": errors}


@router.get("/{employee_id}")
def get_employee(request: Request, employee_id: str):
    employees, salaries, today = _deps(request)
    employee = employees.get_by_id(employee_id)
    current = salaries.get_current(employee.id, today())
    return employee_json(employee, current)


@router.patch("/{employee_id}")
def patch_employee(request: Request, employee_id: str, body: UpdateEmployeeBody):
    employees, salaries, today = _deps(request)
    employee = employees.update_by_id(
        employee_id,
        name=body.name,
        email=body.email,
        country_code=body.country_code,
        department=body.department,
        designation=body.designation,
        status=body.status,  # type: ignore[arg-type]
    )
    current = salaries.get_current(employee.id, today())
    return employee_json(employee, current)


@router.get("/{employee_id}/salary-history")
def salary_history(request: Request, employee_id: str):
    _, salaries, _ = _deps(request)
    return [salary_json(row) for row in salaries.list_history(employee_id)]


@router.post("/{employee_id}/salary", status_code=201)
def add_salary(request: Request, employee_id: str, body: SalaryBody):
    _, salaries, _ = _deps(request)
    salary = salaries.add_salary(
        employee_id=employee_id,
        base_amount=parse_amount(body.base_amount),
        currency=body.currency,
        effective_from=body.effective_from,
        reason=body.reason,
    )
    return salary_json(salary)
